File: backend/app/static/js/usuarios.js
/* ==========================================================================
   ABM de usuarios.

   Todo el estado es local; los datos salen siempre de /api/v1/usuarios.
   El filtrado, la paginación y las reglas de negocio están en el backend
   (Principios 1 y 5): acá no se filtra nada sobre datos ya cargados.
   ========================================================================== */

/* Abreviaturas de mes fijas y no `toLocaleDateString('es-AR', {month:'short'})`:
   ese resultado depende del locale del navegador (un Chrome en inglés daría
   "Oct", y algunas implementaciones agregan punto: "oct."). Con la constante
   la columna se ve igual en cualquier navegador. */
const MESES = ['ene', 'feb', 'mar', 'abr', 'may', 'jun',
               'jul', 'ago', 'sep', 'oct', 'nov', 'dic'];

function abmUsuarios() {
    return {
        usuarios: [],
        roles: [],
        rolesAsignables: [],
        // Puntos de venta activos, de cualquier tipo, para el selector.
        puntosDeVentaAsignables: [],
        total: 0,
        pagina: 1,
        tamano: 10,
        cargando: false,

        // Los tres filtros del diseño de Figma. El backend acepta además
        // username, email y activo: se pueden reactivar sin tocar la API.
        // `activo: 'true'` = solo activos por defecto (ver components/switch_activos).
        filtros: { nombre: '', rol_id: '', local_asignado_id: '', activo: 'true' },

        detalle: { abierto: false, usuario: null },

        form: {
            abierto: false,
            guardando: false,
            id: null,
            nombre: '',
            username: '',
            email: '',
            rol_id: '',
            password: '',
            activo: true,
            fecha_nacimiento: '',
            celular: '',
            local_asignado_id: '',
        },

        // Sección "Accesos permitidos": catálogo plano de permisos, con lo
        // heredado del perfil y lo que se agrega a este usuario.
        accesos: [],
        cargandoAccesos: false,

        confirmacion: { abierta: false, titulo: '', mensaje: '', accion: () => {} },

        async cargar() {
            this.cargando = true;
            try {
                // Los filtros vacíos no se mandan: la API los interpreta
                // como "sin filtrar".
                const params = new URLSearchParams();
                for (const [clave, valor] of Object.entries(this.filtros)) {
                    if (valor !== '' && valor !== null) params.set(clave, valor);
                }

                params.set('pagina', this.pagina);
                params.set('tamano', this.tamano);

                const resp = await fetch('/api/v1/usuarios?' + params, {
                    credentials: 'same-origin',
                });
                if (!resp.ok) throw new Error('No se pudo cargar el listado');

                const datos = await resp.json();
                this.usuarios = datos.resultados;
                this.total = datos.total;
            } catch (e) {
                window.toast(e.message, 'error');
            } finally {
                this.cargando = false;
            }
        },

        async cargarRoles() {
            const resp = await fetch('/api/v1/usuarios/roles-asignables', {
                credentials: 'same-origin',
            });
            if (!resp.ok) return;
            this.rolesAsignables = await resp.json();
            // El filtro por rol muestra los mismos que se pueden asignar.
            this.roles = this.rolesAsignables;
        },

        /**
         * Opciones del selector "Punto de Venta": todos los activos, de
         * cualquier tipo (local, centro de distribución u online).
         *
         * Usa el endpoint del propio módulo y no /puntos-de-venta, que
         * exige permiso de configuración: quien gestiona usuarios no
         * tiene por qué tenerlo.
         */
        async cargarPuntosDeVenta() {
            const resp = await fetch('/api/v1/usuarios/puntos-de-venta-asignables', {
                credentials: 'same-origin',
            });
            if (!resp.ok) return;
            this.puntosDeVentaAsignables = await resp.json();
        },

        limpiar() {
            this.filtros = { nombre: '', rol_id: '', local_asignado_id: '', activo: 'true' };
            this.cargar();
        },

        /* --- Detalle (botón "Ver") --- */

        abrirDetalle(usuario) {
            this.detalle = { abierto: true, usuario };
        },

        // La API devuelve el timestamp crudo: formatearlo es tarea del
        // frontend (Principio 1). Formato del sistema: dd/mm/yy hh:mm.
        formatearFecha(iso) {
            if (!iso) return 'Nunca ingresó';
            return new Date(iso).toLocaleString('es-AR', {
                day: '2-digit', month: '2-digit', year: '2-digit',
                hour: '2-digit', minute: '2-digit', hour12: false,
            });
        },

        /**
         * Fecha sin hora en dd/mm/yyyy, como en el diseño ("06/10/1995").
         *
         * La API manda "1995-10-06" (Principio 1: datos crudos). Se parte
         * el string en lugar de usar new Date(): esa fecha se interpreta
         * como UTC y en Argentina (UTC-3) mostraría el día anterior.
         */
        formatearFechaCorta(iso) {
            if (!iso) return '—';
            const [anio, mes, dia] = iso.split('-');
            return `${dia}/${mes}/${anio}`;
        },

        /**
         * Cumpleaños en dd-mmm ("06-oct"), para la columna del listado.
         *
         * Sin el año a propósito: en la tabla el dato sirve como
         * recordatorio de cumpleaños, no como fecha de nacimiento. El
         * año completo se ve en el panel "Ver".
         *
         * Parte el string por el mismo motivo que formatearFechaCorta:
         * new Date("1995-10-06") se interpreta como UTC y en Argentina
         * mostraría el día anterior.
         */
        formatearCumple(iso) {
            if (!iso) return '—';
            const [, mes, dia] = iso.split('-');
            return `${dia}-${MESES[Number(mes) - 1]}`;
        },

        /* --- Alta y edición --- */

        abrirAlta() {
            this.form = {
                abierto: true, guardando: false, id: null,
                nombre: '', username: '', email: '',
                rol_id: this.rolesAsignables[0]?.id || '', password: '',
                activo: true,
                fecha_nacimiento: '', celular: '', local_asignado_id: '',
            };
            this.cargarAccesos();
        },

        abrirEdicion(usuario) {
            this.form = {
                abierto: true, guardando: false,
                id: usuario.id,
                nombre: usuario.nombre,
                username: usuario.username,
                email: usuario.email || '',
                rol_id: usuario.rol_id,
                password: '',
                activo: usuario.activo,
                // input type="date" espera ISO (yyyy-mm-dd), que es
                // justo lo que devuelve la API.
                fecha_nacimiento: usuario.fecha_nacimiento || '',
                celular: usuario.celular || '',
                local_asignado_id: usuario.local_asignado_id || '',
            };
            this.asegurarPuntoDeVentaActual(usuario);
            this.cargarAccesos();
        },

        /**
         * Mete en el desplegable el punto de venta que el usuario ya tiene,
         * si no está en la lista.
         *
         * Pasa cuando se lo desactivó después de asignárselo: el endpoint
         * solo devuelve los activos, así que sin esto el campo se vería
         * vacío y parecería que el usuario no tiene ninguno. El backend deja
         * conservarlo —lo que prohíbe es asignar uno inactivo—, y el dato ya
         * viene en la respuesta del usuario.
         */
        asegurarPuntoDeVentaActual(usuario) {
            const actual = usuario.local_asignado;
            if (!actual) return;
            if (this.puntosDeVentaAsignables.some((p) => p.id === actual.id)) return;
            this.puntosDeVentaAsignables = [
                ...this.puntosDeVentaAsignables,
                { id: actual.id, nombre: `${actual.nombre} (inactivo)` },
            ];
        },

        /* --- Accesos permitidos --- */

        /**
         * Carga el catálogo de accesos.
         *
         * En edición se piden los del usuario (heredados + individuales).
         * En el alta todavía no hay usuario, así que se piden los del
         * perfil elegido: muestran lo que va a heredar. Por eso también se
         * vuelve a llamar al cambiar el selector de perfil.
         */
        async cargarAccesos() {
            const url = this.form.id
                ? `/api/v1/usuarios/${this.form.id}/accesos`
                : `/api/v1/usuarios/accesos?rol_id=${this.form.rol_id}`;

            if (!this.form.id && !this.form.rol_id) {
                this.accesos = [];
                return;
            }

            // Al cambiar de perfil se conservan las marcas individuales que
            // el usuario ya hizo en el formulario, para no perder su trabajo.
            const marcadosAntes = new Set(
                this.accesos.filter((a) => a.override).map((a) => a.clave)
            );

            this.cargandoAccesos = true;
            try {
                const resp = await fetch(url, { credentials: 'same-origin' });
                if (!resp.ok) throw new Error('No se pudieron cargar los accesos');

                const datos = await resp.json();
                this.accesos = datos.map((a) => ({
                    ...a,
                    override: a.override || marcadosAntes.has(a.clave),
                }));
            } catch (e) {
                window.toast(e.message, 'error');
                this.accesos = [];
            } finally {
                this.cargandoAccesos = false;
            }
        },

        /** Guarda los accesos individuales de un usuario ya existente. */
        async guardarAccesos(usuarioId) {
            const marcados = this.accesos
                .filter((a) => a.override && !a.heredado)
                .map((a) => a.clave);

            const resp = await fetch(`/api/v1/usuarios/${usuarioId}/accesos`, {
                method: 'PUT',
                headers: { 'Content-Type': 'application/json' },
                credentials: 'same-origin',
                body: JSON.stringify({ accesos: marcados }),
            });
            if (!resp.ok) {
                const error = await resp.json().catch(() => ({}));
                throw new Error(error.detail || 'El usuario se guardó, pero fallaron los accesos');
            }
        },

        async guardar() {
            this.form.guardando = true;
            try {
                const esAlta = !this.form.id;
                const cuerpo = {
                    nombre: this.form.nombre,
                    email: this.form.email || null,
                    rol_id: Number(this.form.rol_id),
                    // Los tres campos personales viajan siempre, también
                    // en null: así el backend distingue "vaciar el campo"
                    // de "no lo mandaron" y se pueden borrar desde el form.
                    fecha_nacimiento: this.form.fecha_nacimiento || null,
                    celular: this.form.celular || null,
                    local_asignado_id: this.form.local_asignado_id
                        ? Number(this.form.local_asignado_id)
                        : null,
                };
                if (esAlta) cuerpo.username = this.form.username;
                // En edición, contraseña vacía = no cambiarla.
                if (this.form.password) cuerpo.password = this.form.password;

                const resp = await fetch(
                    esAlta ? '/api/v1/usuarios' : '/api/v1/usuarios/' + this.form.id,
                    {
                        method: esAlta ? 'POST' : 'PUT',
                        headers: { 'Content-Type': 'application/json' },
                        credentials: 'same-origin',
                        body: JSON.stringify(cuerpo),
                    }
                );

                if (!resp.ok) {
                    const error = await resp.json().catch(() => ({}));
                    throw new Error(error.detail || 'No se pudo guardar el usuario');
                }

                // Los accesos se guardan en un segundo paso: en el alta el
                // usuario recién existe (y tiene id) después del POST.
                const guardado = await resp.json();
                await this.guardarAccesos(guardado.id);

                this.form.abierto = false;
                window.toast(esAlta ? 'Usuario creado' : 'Usuario actualizado', 'exito');
                this.cargar();
            } catch (e) {
                window.toast(e.message, 'error');
            } finally {
                this.form.guardando = false;
            }
        },

        /* --- Activar / desactivar, siempre con confirmación --- */

        confirmarEstado(usuario) {
            const desactivando = usuario.activo;
            this.confirmacion = {
                abierta: true,
                titulo: desactivando ? 'Desactivar usuario' : 'Activar usuario',
                mensaje: desactivando
                    ? `¿Desactivar a ${usuario.nombre}? Se cerrarán sus sesiones abiertas y no podrá ingresar. La acción queda auditada.`
                    : `¿Activar a ${usuario.nombre}? Podrá volver a ingresar al sistema.`,
                accion: () => this.cambiarEstado(usuario, !usuario.activo),
            };
        },

        async cambiarEstado(usuario, activo) {
            this.confirmacion.abierta = false;
            try {
                const resp = await fetch(`/api/v1/usuarios/${usuario.id}/estado`, {
                    method: 'PATCH',
                    headers: { 'Content-Type': 'application/json' },
                    credentials: 'same-origin',
                    body: JSON.stringify({ activo }),
                });
                if (!resp.ok) {
                    const error = await resp.json().catch(() => ({}));
                    throw new Error(error.detail || 'No se pudo cambiar el estado');
                }
                window.toast(activo ? 'Usuario activado' : 'Usuario desactivado', 'exito');
                this.cargar();
            } catch (e) {
                window.toast(e.message, 'error');
            }
        },
    };
}

File: backend/app/api/v1/usuarios.py
"""
Endpoints de usuarios.

El acceso al módulo lo decide `requiere_permiso(Modulo.USUARIOS, ...)`.
Sobre qué usuarios concretos puede operar cada rol lo decide el service.
"""

from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query, Request, status
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.core.permisos import Modulo, requiere_permiso
from app.core.utils import ip_de_request
from app.schemas.comunes import RespuestaPaginada
from app.schemas.permisos import (
    AccesoResponse,
    ActualizarAccesosRequest,
    ActualizarPermisosRequest,
    ModuloPermisoEfectivo,
)
from app.schemas.roles import RolResponse
from app.schemas.usuarios import (
    ClaveEspecialResetear,
    ClaveEspecialResultado,
    ClaveEspecialValidar,
    HistorialAccesoResponse,
    LocalResumen,
    UsuarioCrear,
    UsuarioEditar,
    UsuarioEstado,
    UsuarioResponse,
)
from app.services import permisos as servicio_permisos
from app.services import roles as servicio_roles
from app.services import usuarios as servicio_usuarios

router = APIRouter(prefix="/usuarios", tags=["usuarios"])


def _404(exc: Exception):
    return HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))


def _403(exc: Exception):
    return HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(exc))


def _409(exc: Exception):
    return HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


@router.get("", response_model=RespuestaPaginada[UsuarioResponse], summary="Listado de usuarios")
def listar(
    nombre: str | None = Query(default=None),
    username: str | None = Query(default=None),
    email: str | None = Query(default=None),
    rol_id: int | None = Query(default=None),
    local_asignado_id: int | None = Query(default=None),
    activo: bool | None = Query(default=None),
    pagina: int = Query(default=1, ge=1),
    tamano: int = Query(default=50, ge=1, le=200),
    db: Session = Depends(get_db),
    _=Depends(requiere_permiso(Modulo.USUARIOS, "ver")),
):
    """Filtros del Principio 5, todos resueltos en el backend."""
    resultados, total = servicio_usuarios.listar_usuarios(
        db,
        nombre=nombre,
        username=username,
        email=email,
        rol_id=rol_id,
        local_asignado_id=local_asignado_id,
        activo=activo,
        pagina=pagina,
        tamano=tamano,
    )
    return RespuestaPaginada[UsuarioResponse](
        total=total, pagina=pagina, tamano=tamano, resultados=resultados
    )


@router.get("/roles-asignables", response_model=list[RolResponse], summary="Roles que puede asignar")
def roles_asignables(
    db: Session = Depends(get_db),
    autor=Depends(requiere_permiso(Modulo.USUARIOS, "ver")),
):
    """Alimenta el selector de rol del formulario de alta/edición."""
    return servicio_usuarios.roles_asignables(db, autor)


@router.get(
    "/locales-asignables",
    response_model=list[LocalResumen],
    summary="Locales que se pueden asignar",
)
def locales_asignables(
    db: Session = Depends(get_db),
    _=Depends(requiere_permiso(Modulo.USUARIOS, "ver")),
):
    """
    Alimenta el selector "Local Asignado" del formulario.

    Existe acá, y no se reusa GET /puntos-de-venta, porque ese endpoint
    exige permiso de CONFIGURACION: quien gestiona usuarios no tiene por
    qué tenerlo, y se quedaría sin opciones en el desplegable.
    """
    return servicio_usuarios.locales_asignables(db)


@router.get(
    "/accesos",
    response_model=list[AccesoResponse],
    summary="Accesos que otorga un rol",
)
def accesos_de_rol(
    rol_id: int = Query(description="Rol del que se quieren ver los accesos"),
    db: Session = Depends(get_db),
    _=Depends(requiere_permiso(Modulo.USUARIOS, "ver")),
):
    """
    Alimenta la sección "Accesos permitidos" en el ALTA, cuando el usuario
    todavía no existe: muestra lo que va a heredar del rol elegido.
    """
    try:
        servicio_roles.obtener_rol(db, rol_id)
    except servicio_roles.NoEncontrado as exc:
        raise _404(exc) from exc

    return servicio_permisos.accesos_de_rol(db, rol_id)


@router.get(
    "/{usuario_id}/accesos",
    response_model=list[AccesoResponse],
    summary="Accesos efectivos de un usuario",
)
def accesos_de_usuario(
    usuario_id: int,
    db: Session = Depends(get_db),
    _=Depends(requiere_permiso(Modulo.USUARIOS, "ver")),
):
    """Lo mismo que el anterior, pero para un usuario ya existente."""
    try:
        usuario = servicio_usuarios.obtener_usuario(db, usuario_id)
    except servicio_roles.NoEncontrado as exc:
        raise _404(exc) from exc

    return servicio_permisos.accesos_de_usuario(db, usuario)


@router.put(
    "/{usuario_id}/accesos",
    response_model=list[AccesoResponse],
    summary="Guardar accesos individuales",
)
def actualizar_accesos(
    usuario_id: int,
    datos: ActualizarAccesosRequest,
    request: Request,
    db: Session = Depends(get_db),
    autor=Depends(requiere_permiso(Modulo.USUARIOS, "editar")),
):
    """
    Guarda los accesos marcados como overrides individuales.

    Solo toca la acción de cada acceso del catálogo: los permisos que el
    usuario tenga cargados desde la pantalla completa quedan intactos.
    """
    try:
        usuario = servicio_usuarios.obtener_usuario(db, usuario_id)
        servicio_usuarios.validar_puede_gestionar(autor, usuario.rol)
        accesos = servicio_permisos.actualizar_accesos_usuario(
            db, usuario, datos.accesos, autor.id, ip_de_request(request)
        )
    except servicio_usuarios.SinPermiso as exc:
        raise _403(exc) from exc
    except servicio_roles.NoEncontrado as exc:
        raise _404(exc) from exc

    db.commit()
    return accesos


@router.get("/{usuario_id}", response_model=UsuarioResponse, summary="Detalle de usuario")
def detalle(
    usuario_id: int,
    db: Session = Depends(get_db),
    _=Depends(requiere_permiso(Modulo.USUARIOS, "ver")),
):
    try:
        return servicio_usuarios.obtener_usuario(db, usuario_id)
    except servicio_roles.NoEncontrado as exc:
        raise _404(exc) from exc


@router.post("", response_model=UsuarioResponse, status_code=status.HTTP_201_CREATED, summary="Alta de usuario")
def crear(
    datos: UsuarioCrear,
    request: Request,
    db: Session = Depends(get_db),
    autor=Depends(requiere_permiso(Modulo.USUARIOS, "crear")),
):
    try:
        usuario = servicio_usuarios.crear_usuario(
            db,
            autor,
            username=datos.username,
            nombre=datos.nombre,
            password=datos.password,
            rol_id=datos.rol_id,
            email=datos.email,
            fecha_nacimiento=datos.fecha_nacimiento,
            celular=datos.celular,
            local_asignado_id=datos.local_asignado_id,
            ip_origen=ip_de_request(request),
        )
    except servicio_usuarios.SinPermiso as exc:
        raise _403(exc) from exc
    except servicio_roles.NoEncontrado as exc:
        raise _404(exc) from exc
    except servicio_roles.ReglaDeNegocio as exc:
        raise _409(exc) from exc

    db.commit()
    return usuario


@router.put("/{usuario_id}", response_model=UsuarioResponse, summary="Editar usuario")
def editar(
    usuario_id: int,
    datos: UsuarioEditar,
    request: Request,
    db: Session = Depends(get_db),
    autor=Depends(requiere_permiso(Modulo.USUARIOS, "editar")),
):
    try:
        usuario = servicio_usuarios.editar_usuario(
            db,
            autor,
            usuario_id,
            nombre=datos.nombre,
            email=datos.email,
            rol_id=datos.rol_id,
            password=datos.password,
            fecha_nacimiento=datos.fecha_nacimiento,
            celular=datos.celular,
            local_asignado_id=datos.local_asignado_id,
            # Distinguen "no lo mandaron" de "lo mandaron vacío": los tres
            # campos son opcionales y se tienen que poder borrar.
            editar_fecha_nacimiento="fecha_nacimiento" in datos.model_fields_set,
            editar_celular="celular" in datos.model_fields_set,
            editar_local="local_asignado_id" in datos.model_fields_set,
            ip_origen=ip_de_request(request),
        )
    except servicio_usuarios.SinPermiso as exc:
        raise _403(exc) from exc
    except servicio_roles.NoEncontrado as exc:
        raise _404(exc) from exc
    except servicio_roles.ReglaDeNegocio as exc:
        raise _409(exc) from exc

    db.commit()
    return usuario


@router.patch("/{usuario_id}/estado", response_model=UsuarioResponse, summary="Activar o desactivar")
def cambiar_estado(
    usuario_id: int,
    datos: UsuarioEstado,
    request: Request,
    db: Session = Depends(get_db),
    autor=Depends(requiere_permiso(Modulo.USUARIOS, "editar")),
):
    try:
        usuario = servicio_usuarios.cambiar_estado_usuario(
            db, autor, usuario_id, datos.activo, ip_de_request(request)
        )
    except servicio_usuarios.SinPermiso as exc:
        raise _403(exc) from exc
    except servicio_roles.NoEncontrado as exc:
        raise _404(exc) from exc
    except servicio_roles.ReglaDeNegocio as exc:
        raise _409(exc) from exc

    db.commit()
    return usuario


@router.get(
    "/{usuario_id}/permisos",
    response_model=list[ModuloPermisoEfectivo],
    summary="Árbol de permisos efectivos",
)
def obtener_permisos(
    usuario_id: int,
    db: Session = Depends(get_db),
    _=Depends(requiere_permiso(Modulo.USUARIOS, "ver")),
):
    """
    Devuelve el permiso efectivo (rol OR override) y, por separado, qué
    parte viene heredada del rol y qué parte es override individual.
    """
    try:
        usuario = servicio_usuarios.obtener_usuario(db, usuario_id)
    except servicio_roles.NoEncontrado as exc:
        raise _404(exc) from exc

    return servicio_permisos.arbol_de_usuario(db, usuario)


@router.put(
    "/{usuario_id}/permisos",
    response_model=list[ModuloPermisoEfectivo],
    summary="Actualizar overrides de permisos",
)
def actualizar_permisos(
    usuario_id: int,
    datos: ActualizarPermisosRequest,
    request: Request,
    db: Session = Depends(get_db),
    autor=Depends(requiere_permiso(Modulo.USUARIOS, "editar")),
):
    try:
        usuario = servicio_usuarios.obtener_usuario(db, usuario_id)
        servicio_usuarios.validar_puede_gestionar(autor, usuario.rol)
        arbol = servicio_permisos.actualizar_permisos_usuario(
            db,
            usuario,
            [p.model_dump() for p in datos.permisos],
            autor.id,
            ip_de_request(request),
        )
    except servicio_usuarios.SinPermiso as exc:
        raise _403(exc) from exc
    except servicio_roles.NoEncontrado as exc:
        raise _404(exc) from exc
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc)) from exc

    db.commit()
    return arbol


@router.get(
    "/{usuario_id}/historial",
    response_model=RespuestaPaginada[HistorialAccesoResponse],
    summary="Historial de accesos",
)
def historial(
    usuario_id: int,
    desde: date | None = Query(default=None),
    hasta: date | None = Query(default=None),
    resultado: str | None = Query(default=None, pattern="^(exitoso|fallido)$"),
    pagina: int = Query(default=1, ge=1),
    tamano: int = Query(default=50, ge=1, le=200),
    db: Session = Depends(get_db),
    _=Depends(requiere_permiso(Modulo.USUARIOS, "ver")),
):
    try:
        filas, total = servicio_usuarios.historial_de_usuario(
            db, usuario_id, desde=desde, hasta=hasta, resultado=resultado,
            pagina=pagina, tamano=tamano,
        )
    except servicio_roles.NoEncontrado as exc:
        raise _404(exc) from exc

    return RespuestaPaginada[HistorialAccesoResponse](
        total=total, pagina=pagina, tamano=tamano, resultados=filas
    )


# ----------------------------------------------------------------------------
# Clave especial — solo la Cuenta Maestra tiene una.
# Para cualquier otro usuario estos endpoints devuelven 404.
# ----------------------------------------------------------------------------


@router.post(
    "/{usuario_id}/clave-especial/validar",
    response_model=ClaveEspecialResultado,
    summary="Validar clave especial",
)
def validar_clave_especial(
    usuario_id: int,
    datos: ClaveEspecialValidar,
    db: Session = Depends(get_db),
    _=Depends(requiere_permiso(Modulo.USUARIOS, "ver")),
):
    try:
        valida = servicio_usuarios.validar_clave_especial(db, usuario_id, datos.clave)
    except servicio_roles.NoEncontrado as exc:
        raise _404(exc) from exc

    return ClaveEspecialResultado(valida=valida)


@router.patch(
    "/{usuario_id}/clave-especial/resetear",
    response_model=ClaveEspecialResultado,
    summary="Resetear clave especial",
)
def resetear_clave_especial(
    usuario_id: int,
    datos: ClaveEspecialResetear,
    request: Request,
    db: Session = Depends(get_db),
    autor=Depends(requiere_permiso(Modulo.USUARIOS, "editar")),
):
    try:
        servicio_usuarios.resetear_clave_especial(
            db, autor, usuario_id, datos.clave_nueva, ip_de_request(request)
        )
    except servicio_roles.NoEncontrado as exc:
        raise _404(exc) from exc

    db.commit()
    return ClaveEspecialResultado(valida=True)
